Support Circles

ArtificialOwl opened the pull request:
This PR will add support to a non-admin group shares.

It includes:
- call to the Circles Sharees API to search for Circles when the user want to share a file/folder.
- call to the Circles Share Provider when browsing files (as file owner and as shared user) and editing rights/permissions.

~~WARNING: There is a known bug with the resharing of the content of a Circle. This need to be fixed !~~ (fixed)

Note: I know, I haven't signed any push ...

ArtificialOwl committed:
rework on getAllProviders()

ArtificialOwl committed:
Circles

Signed-off-by: Maxence Lange maxence@nextcloud.com

ArtificialOwl committed:
getSharedByCircleProvider()

ArtificialOwl committed:
share_type_circle (=7)

ArtificialOwl committed:
check that the circle exists

ArtificialOwl committed:
get circles on search

ArtificialOwl committed:
exact result

ArtificialOwl committed:
display sharing out of right panel

ArtificialOwl committed:
getSharesBy on TYPE_CIRCLE

ArtificialOwl committed:
renaming method

ArtificialOwl committed:
display share

ArtificialOwl committed:
get circle share provider

ArtificialOwl committed:
fix display name

ArtificialOwl committed:
key_exists

ArtificialOwl committed:
if not, getCache() runs on null

ArtificialOwl committed:
SharedWithMe()

ArtificialOwl committed:
display issue

Signed-off-by: Maxence Lange maxence@nextcloud.com

MorrisJobke commented on apps/files_sharing/lib/Controller/ShareAPIController.php:
```diff
@@ -592,7 +608,14 @@ public function getShares(
 			$mailShares = [];
 		}
-		$shares = array_merge($userShares, $groupShares, $linkShares, $mailShares);
+//		if (\OCP\App::isEnabled('circles')) {
```
Remove it 😉

MorrisJobke commented on apps/files_sharing/lib/Controller/ShareesAPIController.php:
```diff
@@ -294,6 +296,19 @@ protected function getGroups($search) {
+		$this->result['circles'] = $this->result['exact']['circles'] = [];
+
+		$result = \OCA\Circles\Api\Sharees::search($search, $this->limit, $this->offset);
+		if (array_key_exists('circles', $result['exact'])) $this->result['exact']['circles'] = $result['exact']['circles'];
```
Always use `{` and `}` for `if` and a new line for the stuff inside of the block

MorrisJobke commented on apps/files_sharing/tests/js/sharedbreadcrumviewSpec.js:
```diff
@@ -120,6 +120,21 @@ describe('OCA.Sharing.ShareBreadCrumbView tests', function() {
 			expect(bc.$el.find('.shared').length).toEqual(1);
 			expect(bc.$el.find('.icon-public').length).toEqual(1);
 		});
+		it('Render shared if dir is shared by link', function() {
```
`... is shared by circle`

MorrisJobke commented:
> Note: I know, I haven't signed any push ...

Just squash all of those commits into one and sign it then 😉

ArtificialOwl committed:
Merge remote-tracking branch 'origin/shares-circles' into shares-circles

Signed-off-by: Maxence Lange maxence@nextcloud.com

daita commented:
@MorrisJobke I fixed the code style

rullzer committed:
We need to verify access to the share

With user/group shares and all we need to check if the circle share is
accessible. For now I assume that if it is a circle share it is
accessible. This might (probably) is incrorrect.

It would make sense I guess to move this over to the providers. But for
now this has to do.

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer commented on apps/files_sharing/lib/Controller/ShareAPIController.php:
```diff
@@ -443,6 +447,16 @@ public function createShare(
 			$share->setSharedWith($shareWith);
+		} else if ($shareType === \OCP\Share::SHARE_TYPE_CIRCLE) {
+
+			$circle = \OCA\Circles\Api\Circles::detailsCircle($shareWith);
```
This will fail hard if the app is not installed... Of course then there probabaly won't be a circle share. But if the app gets disabled or so. It will most likely do :boom:

daita replied:
it is in createShare, so I doubt anyone could create a share to a Circle if the app is not enabled, but I will add a condition on `(\OCP\App::isEnabled('circles'))`

rullzer commented on apps/files_sharing/lib/SharedStorage.php:
```diff
@@ -97,6 +97,8 @@ public function __construct($arguments) {
 	 * @return ICacheEntry
 	 */
 	private function getSourceRootInfo() {
+		$this->init();
```
@icewind1991 please comment.

icewind1991 replied:
👎 will have a massive impact on performance

daita replied:
even with the check on the value ?
> if ($this->initialized) { } 

The issue I had (if I remember well) is that superStorage was null

ArtificialOwl committed:
Check if the app is enabled before creating a new share

Signed-off-by: Maxence Lange <maxence@nextcloud.com>

ArtificialOwl committed:
Merge remote-tracking branch 'origin/shares-circles' into shares-circles

ArtificialOwl committed:
mistake were made

Signed-off-by: Maxence Lange <maxence@nextcloud.com>

ArtificialOwl committed:
not needed anymore

Signed-off-by: Maxence Lange <maxence@nextcloud.com>

ArtificialOwl committed:
display the owner of the circle

useful to catch resharing

Signed-off-by: Maxence Lange <maxence@nextcloud.com>

daita commented:
@MorrisJobke @rullzer @icewind1991  Fixed the few stuff you were asking

ArtificialOwl committed:
useless line

Signed-off-by: Maxence Lange <maxence@nextcloud.com>

ArtificialOwl committed:
Merge remote-tracking branch 'origin/master' into shares-circles

ArtificialOwl committed:
few tests

Signed-off-by: Maxence Lange <maxence@nextcloud.com>

ArtificialOwl committed:
creating an empty share-merge test

Signed-off-by: Maxence Lange <maxence@nextcloud.com>

ArtificialOwl committed:
few more tests

Signed-off-by: Maxence Lange <maxence@nextcloud.com>

ArtificialOwl committed:
few more tests

Signed-off-by: Maxence Lange <maxence@nextcloud.com>

MorrisJobke commented:
> Note: I know, I haven't signed any push ...

Squash all commits, sign off this commit and then I would say this can get in :)

MorrisJobke commented:
Let's continue in #3906 - next time just force push to the same branch to overwrite the github version and update the GitHub PR 😉
